tests/client/core/components/TestPaginatorLink.js

```diff
@@ -0,0 +1,121 @@
+
+import PaginatorLink from 'core/components/PaginatorLink';
+
+describe('<PaginatorLink />', () => {
+  const renderProps = Object.freeze({
+    currentPage: 2,
+    page: 3,
+    pageCount: 4,
+    pathname: '/some/link',
+  });
+
+  function renderLink(customProps = {}) {
+    const props = { ...renderProps, ...customProps };
+    return findDOMNode(
+      renderIntoDocument(<PaginatorLink {...props} />));
+  }
+
+  it('requires currentPage', () => {
+    const props = { ...renderProps };
+    delete props.currentPage;
+    assert.throws(
+      () => renderIntoDocument(<PaginatorLink {...props} />),
+      /currentPage .* cannot be undefined/);
+  });
+
+  it('requires pathname', () => {
+    const props = { ...renderProps };
+    delete props.pathname;
+    assert.throws(
+      () => renderIntoDocument(<PaginatorLink {...props} />),
+      /pathname .* cannot be undefined/);
+  });
+
+  it('requires a page', () => {
+    const props = { ...renderProps };
+    delete props.page;
+    assert.throws(
+      () => renderIntoDocument(<PaginatorLink {...props} />),
+      /page .* cannot be undefined/);
+  });
+
+  it('requires pageCount', () => {
+    const props = { ...renderProps };
+    delete props.pageCount;
+    assert.throws(
+      () => renderIntoDocument(<PaginatorLink {...props} />),
+      /pageCount .* cannot be undefined/);
+  });
+
+  describe('when the link is to the current page', () => {
+    it('does not contain a link', () => {
+      const item = renderLink({ currentPage: 3, page: 3 });
+      assert.strictEqual(item.querySelector('a'), null);
+      assert.equal(item.textContent, '3');
+    });
+
+    it('uses the provided text', () => {
+      const item = renderLink({
+        currentPage: 3, page: 3, text: 'go to page',
+      });
+      assert.equal(item.tagName, 'SPAN');
+      assert.equal(item.textContent, 'go to page');
+    });
+  });
+
+  describe('when the link is to a different page', () => {
+    it('has a link', () => {
+      const link = renderLink({ page: 3 });
+      assert.equal(link.tagName, 'A');
+      assert.equal(link.textContent, '3');
+    });
+
+    it('uses the provided text', () => {
+      const link = renderLink({ page: 3, text: 'go to next page' });
+      assert.equal(link.tagName, 'A');
+      assert.equal(link.textContent, 'go to next page');
+    });
+
+    it('renders a custom link component', () => {
+      const LinkComponent = sinon.spy(() => <div />);
+      renderLink({ page: 3, LinkComponent });
+      assert.ok(LinkComponent.called, 'custom LinkComponent was not rendered');
+    });
+
+    it('passes query params to the link', () => {
+      const page = 3;
+      const queryParams = { color: 'red' };
+      const LinkComponent = sinon.spy(() => <div />);
+      renderLink({ queryParams, page, LinkComponent });
+
+      assert.ok(LinkComponent.called, 'custom LinkComponent was not rendered');
+      const props = LinkComponent.firstCall.args[0];
```

Comment:
Ah, yeah, I agree about doing setup first and assertions later. However, the `props` assignment can't be moved up because it relies on the stub getting called correctly. If I move it up above `assert.ok(stub.called)` then the test fails in a harder to diagnose error in case it wasn't called:
````
LinkComponent.firstCall is null
````